// react-tackle-box/lib/Slot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.default = createSlot;
exports.SlotProvider = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/builtin/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/builtin/inheritsLoose"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var listener = null;
var initialSlots = [];
var slots = [];

var SlotProvider =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(SlotProvider, _React$Component);

  function SlotProvider() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    if (listener) console.warn('Oops more than one Slot Provider'); //eslint-disable-line no-console

    if (initialSlots.length) initialSlots.map(function (slot) {
      return _this.addSlot(slot);
    });
    initialSlots.length = 0;
    var timer;

    listener = function listener(slot) {
      _this.addSlot(slot);

      clearTimeout(timer);
      timer = setTimeout(function () {
        return _this.forceUpdate();
      });
    };

    return _this;
  }

  var _proto = SlotProvider.prototype;

  _proto.addSlot = function addSlot(Provider) {
    var _this2 = this;

    var idx = slots.length;
    slots.push({
      Provider: Provider,
      context: {
        anchor: null,
        set: function set(anchor) {
          var _slots$idx = slots[idx],
              Provider = _slots$idx.Provider,
              context = _slots$idx.context;
          slots[idx] = {
            Provider: Provider,
            context: {
              anchor: anchor,
              set: context.set
            }
          };

          _this2.forceUpdate();
        }
      }
    });
  };

  _proto.render = function render() {
    var children = this.props.children;
    return slots.reduce(function (child, _ref) {
      var Provider = _ref.Provider,
          context = _ref.context;
      return _react.default.createElement(Provider, {
        value: context
      }, child || children);
    }, null);
  };

  return SlotProvider;
}(_react.default.Component);

exports.SlotProvider = SlotProvider;

function createSlot() {
  var _React$createContext = _react.default.createContext({
    anchor: null,
    set: null
  }),
      Provider = _React$createContext.Provider,
      Consumer = _React$createContext.Consumer;

  var Outlet =
  /*#__PURE__*/
  function (_React$Component2) {
    (0, _inheritsLoose2.default)(Outlet, _React$Component2);

    function Outlet() {
      return _React$Component2.apply(this, arguments) || this;
    }

    var _proto2 = Outlet.prototype;

    _proto2.shouldComponentUpdate = function shouldComponentUpdate() {
      return false;
    };

    _proto2.render = function render() {
      var _this3 = this;

      return _react.default.createElement(Consumer, null, function (_ref2) {
        var set = _ref2.set;
        return _react.default.createElement("div", (0, _extends2.default)({
          ref: set
        }, _this3.props));
      });
    };

    return Outlet;
  }(_react.default.Component);

  var Entry = _react.default.forwardRef(function (_ref3) {
    var children = _ref3.children,
        waitForOutlet = _ref3.waitForOutlet;
    return _react.default.createElement(Consumer, null, function (_ref4) {
      var anchor = _ref4.anchor;
      var child = typeof children === 'function' ? children(!anchor) : children;
      if (anchor) return _reactDom.default.createPortal(child, anchor);
      return !waitForOutlet ? child : null;
    });
  });

  listener ? listener(Provider) : initialSlots.push(Provider);
  return {
    Entry: Entry,
    Outlet: Outlet
  };
}
