// @4c/layout/lib/PropTypes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/builtin/interopRequireDefault");

exports.__esModule = true;
exports.defaultProps = exports.propTypes = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _elementType = _interopRequireDefault(require("prop-types-extra/lib/elementType"));

const propTypes = {
  direction: _propTypes.default.oneOf(['row', 'column']).isRequired,
  as: _elementType.default.isRequired,
  pad: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.number]),
  wrap: _propTypes.default.bool,
  grow: _propTypes.default.bool,
  inline: _propTypes.default.bool,
  flex: _propTypes.default.any,
  alignSelf: _propTypes.default.oneOf(['flex-start', 'flex-end', 'center', 'stretch']),
  align: _propTypes.default.oneOf(['flex-start', 'flex-end', 'center', 'stretch', 'baseline']).isRequired,
  justify: _propTypes.default.oneOf(['flex-start', 'flex-end', 'center', 'baseline', 'space-between', 'space-around']).isRequired
};
exports.propTypes = propTypes;
const defaultProps = {
  as: 'div',
  direction: 'row',
  align: 'stretch',
  justify: 'flex-start'
};
exports.defaultProps = defaultProps;
